core, trie: new trie

fjl opened the pull request:
This PR contains a rewrite of package trie with the following goals:

- reduced API surface: unexported node types, string funcs, other internals
- documented core internals
- documented API
- improved performance
- more memory reuse, especially when hashing the trie
- hashing is independent of the database

Gustav-Simonsson commented on trie/trie.go:
```diff
@@ -2,47 +2,47 @@ package trie
 
-func ParanoiaCheck(t1 *Trie, backend Backend) (bool, *Trie) {
-	t2 := New(nil, backend)
+const lruSize = 200
```
What unit is this, number of nodes, leaf values?

fjl replied:
Number of DB entries.

Gustav-Simonsson commented on trie/trie.go:
```diff
@@ -2,47 +2,47 @@ package trie
+		if backend == nil {
+			panic("cannot use exisiting root without a database")
+		}
+		// if v, _ := trie.db.Get(root[:]); len(v) == 0 {
```
Can this be removed?

fjl replied:
I added this for debugging. Please note that the PR is not done yet.If it is uncommented, several tests panic.

Gustav-Simonsson commented on trie/trie.go:
```diff
@@ -50,230 +50,166 @@ func (self *Trie) Iterator() *Iterator {
+}
 
-	self.cache.Reset()
+var emptyRoot = common.HexToHash("56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")
```
Please add a comment explaining this value.

Gustav-Simonsson commented on trie/trie.go:
```diff
@@ -50,230 +50,166 @@ func (self *Trie) Iterator() *Iterator {
 	self.mu.Lock()
 	defer self.mu.Unlock()
+	self.root = self.hashRoot(self.db)
+	// self.db.Flush()
```
If it's commented out for a reason would be good to document.

Gustav-Simonsson commented on trie/trie.go:
```diff
@@ -50,230 +50,166 @@ func (self *Trie) Iterator() *Iterator {
-	default:
-		panic(fmt.Sprintf("%T: invalid node: %v", node, node))
-	}
+func concat(s1 []byte, s2 ...byte) []byte {
```
I believe you've re-invented append.
```go
append(s S, x ...T) S  // T is the element type of S
```

fjl replied:
If only there was a way to indicate that this PR is in progress. Maybe a label of some sort. Someone should invent that. append‎, being nice and all, is allowed to reuse the backing array of its first argument if it has sufficient cap. concat doesn't reuse the first argument but returns a fresh slice every time. I'll leave it as an exercise for you to puzzle about why concat is being used instead of append in the one place where it is used. 

Gustav-Simonsson replied:
I'll puzzle that it's used instead of append to ensure there is always a newly allocated byte slice :) Thanks for explaining!

fjl commented:
Preliminary performance report:

```
benchmark                                       old ns/op     new ns/op     delta
BenchmarkInsertChain_empty_memdb-8              109218        64479         -40.96%
BenchmarkInsertChain_empty_diskdb-8             194874        136614        -29.90%
BenchmarkInsertChain_valueTx_memdb-8            240241        156505        -34.85%
BenchmarkInsertChain_valueTx_diskdb-8           412044        322234        -21.80%
BenchmarkInsertChain_valueTx_100kB_memdb-8      1203981       1010477       -16.07%
BenchmarkInsertChain_valueTx_100kB_diskdb-8     3067184       2969643       -3.18%
BenchmarkInsertChain_uncles_memdb-8             213715        170358        -20.29%
BenchmarkInsertChain_uncles_diskdb-8            301778        251242        -16.75%
BenchmarkInsertChain_ring200_memdb-8            144237467     81842282      -43.26%
BenchmarkInsertChain_ring200_diskdb-8           170420433     93856663      -44.93%
BenchmarkInsertChain_ring1000_memdb-8           194525358     116581937     -40.07%
BenchmarkInsertChain_ring1000_diskdb-8          219006749     127585191     -41.74%
BenchmarkCompactEncode                          135           61.4          -54.52%
BenchmarkCompactDecode                          87.8          87.2          -0.68%
BenchmarkCompactHexDecode                       116           113           -2.59%
BenchmarkDecodeCompact                          53.8          53.4          -0.74%
BenchmarkGet                                    628           289           -53.98%
BenchmarkGetDB                                  518           1471          +183.98%
BenchmarkUpdateBE                               2935          2002          -31.79%
BenchmarkUpdateLE                               3289          2409          -26.76%
BenchmarkHashBE                                 100965203     40623083      -59.77%
BenchmarkHashLE                                 117250940     43654768      -62.77%

benchmark                                       old allocs     new allocs     delta
BenchmarkInsertChain_empty_memdb-8              743            330            -55.59%
BenchmarkInsertChain_empty_diskdb-8             810            388            -52.10%
BenchmarkInsertChain_valueTx_memdb-8            1500           790            -47.33%
BenchmarkInsertChain_valueTx_diskdb-8           1616           890            -44.93%
BenchmarkInsertChain_valueTx_100kB_memdb-8      1504           794            -47.21%
BenchmarkInsertChain_valueTx_100kB_diskdb-8     1847           1029           -44.29%
BenchmarkInsertChain_uncles_memdb-8             1195           782            -34.56%
BenchmarkInsertChain_uncles_diskdb-8            1260           846            -32.86%
BenchmarkInsertChain_ring200_memdb-8            1182124        339278         -71.30%
BenchmarkInsertChain_ring200_diskdb-8           1220860        345026         -71.74%
BenchmarkInsertChain_ring1000_memdb-8           1600303        429111         -73.19%
BenchmarkInsertChain_ring1000_diskdb-8          1638553        435461         -73.42%
BenchmarkCompactEncode                          2              1              -50.00%
BenchmarkCompactDecode                          1              1              +0.00%
BenchmarkCompactHexDecode                       1              1              +0.00%
BenchmarkDecodeCompact                          1              1              +0.00%
BenchmarkGet                                    2              1              -50.00%
BenchmarkGetDB                                  2              6              +200.00%
BenchmarkUpdateBE                               14             8              -42.86%
BenchmarkUpdateLE                               13             8              -38.46%
BenchmarkHashBE                                 392188         106708         -72.79%
BenchmarkHashLE                                 473505         111063         -76.54%

benchmark                                       old bytes     new bytes     delta
BenchmarkInsertChain_empty_memdb-8              46288         19827         -57.17%
BenchmarkInsertChain_empty_diskdb-8             54900         28002         -48.99%
BenchmarkInsertChain_valueTx_memdb-8            100512        48822         -51.43%
BenchmarkInsertChain_valueTx_diskdb-8           115107        62805         -45.44%
BenchmarkInsertChain_valueTx_100kB_memdb-8      1376513       899592        -34.65%
BenchmarkInsertChain_valueTx_100kB_diskdb-8     3008558       2297909       -23.62%
BenchmarkInsertChain_uncles_memdb-8             68362         41908         -38.70%
BenchmarkInsertChain_uncles_diskdb-8            79106         55894         -29.34%
BenchmarkInsertChain_ring200_memdb-8            94582187      18819301      -80.10%
BenchmarkInsertChain_ring200_diskdb-8           96704176      20032446      -79.28%
BenchmarkInsertChain_ring1000_memdb-8           137721565     23260916      -83.11%
BenchmarkInsertChain_ring1000_diskdb-8          139818770     24621324      -82.39%
BenchmarkCompactEncode                          16            4             -75.00%
BenchmarkCompactDecode                          16            16            +0.00%
BenchmarkCompactHexDecode                       32            32            +0.00%
BenchmarkDecodeCompact                          4             4             +0.00%
BenchmarkGet                                    144           80            -44.44%
BenchmarkGetDB                                  144           320           +122.22%
BenchmarkUpdateBE                               1766          1567          -11.27%
BenchmarkUpdateLE                               1797          1627          -9.46%
BenchmarkHashBE                                 223376483     3735414       -98.33%
BenchmarkHashLE                                 244950514     3927063       -98.40%
```

zelig commented on trie/secure_trie.go:
```diff
@@ -16,46 +16,84 @@
+//
+// If root is the zero hash or the sha3 hash of an empty string, the
+// trie is initially empty. Otherwise, New will panics if db is nil or
+// root does not exist in the database. Accessing the trie loads nodes
```
will panic [typo]

fjl replied:
It will return an error in the final version of this PR.

zelig commented:
is ethdb.Database put/get meant to be concurrently used? cos AFAIU LDBDatabase implementation is threadsafe due to leveldb and batch calls being safe but MemDatabase is not, nor is its batch.

fjl commented:
Trie is not thread safe either.

zelig commented:
Yes the trie is clearly documented not to be. But ethdb.Database is independent and ldbd is actually perfectly functional concurrently right? Either way it's worth a comment

fjl commented:
I've rebased this and added my initial stab at the merkle proof functions. Some comments on those below. My plan is to move parts of this (rlp changes #1778, ethdb changes) into develop soon-ish and wait for the other pending core changes to be merged before bringing in the trie. This also leaves more time for reviewing.

About the merkle proof code: @ebuchman did an implementation for the old trie in #1624, but I couldn't port these directly because the API wasn't quite right. For the light client, the initial use of these proofs is that we have the state root hash and wish to retrieve the proof and value for a key in one request. The API in #1624 expected the value to be passed as an input to `Verify`. In addition, the PR defined a custom encoding for proofs. AFAIK the plan is to represent proofs [as a list of encoded nodes as defined in the YP](https://github.com/ethereum/wiki/wiki/Ethereum-Wire-Protocol#proposal-pv64).

@zsfelfoldi Please comment on the API. The two functions of interest are

```
func (*Trie) Prove(key []byte) []rlp.RawValue
func VerifyProof(root common.Hash, key []byte, proof []rlp.RawValue) (value []byte, err error)
```

zsfelfoldi commented:
The prove/verify API is perfect for me, I'll probably add support for my skipLevels improvement (omitting the highest levels from the proof if we already have them) later if there's a consensus about adding it to the eth/64 protocol in the first place. I agree about the value being a result of verify, not an input parameter (although it can be read from the proof itself but it's illogical to pass as an input).

zsfelfoldi commented:
You can also leave the debugging of proofs to me. Do you think there's a chance that the new trie can be merged into develop in the next few days? I need it but I'm already rebased on top of Peter's #1779 and I really wouldn't like to go any further away from develop :)

ebuchman commented:
Nice work! I agree the Value doesn't belong in the API, I was using it for sanity checking while testing and forgot to remove (it only just checks that the value passed in and the value at the bottom of the proof are the same). Also wasn't sure about encoding so just tried something simple. But looks like you have a shorter implementation anyways so great! I'll take a look at some testing.

fjl commented:
@zsfelfoldi rebased on develop
